### PR TITLE
GRPC Streams Metrics

### PR DESCRIPTION
**Pull request by olegbespalov**
# What?

This brings a few GRPC streams metrics to the module.

image

 :thinking:  Two questions to the reviewers:
* Currently, we also issue the `grpc_req_duration`, which also counts streams. Is that okay?
* Do you have in mind any other metrics that could be useful?

# Why?

Brings new metrics to the GRPC module and...

Resolves: #9

### CODE REVIEW

**Review by codebien — CHANGES_REQUESTED** [review]
> Currently, we also issue the grpc_req_duration, which also counts streams. Is that okay?

I'm fine with it, but what does it include for a stream? Does it contain both or only the sent messages?

>Do you have in mind any other metrics that could be useful?

I would prefer a minimal approach because metrics could be difficult to remove. So I would prefer if people asks to add them specifically. It would be difficult to get a negative feedback around it. No-one will tell us what they do not use, instead, it is easier to get a request to add a new one.

File: grpc/metrics.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package grpc
+
+import "go.k6.io/k6/metrics"
+
+// grpcMetrics contains the metrics for the grpc extension.
+type grpcMetrics struct {
+	GRPCStreams                 *metrics.Metric
+	GRPCStreamsMessagesSent     *metrics.Metric
+	GRPCStreamsMessagesReceived *metrics.Metric
+}
```
Comment (codebien) [inline grpc/metrics.go:10]:
```suggestion
// streamsMetrics contains the metrics for the grpc extension.
type streamsMetrics struct {
	Streams                 *metrics.Metric
	MessagesSent     *metrics.Metric
	MessagesReceived *metrics.Metric
}
```
I'm for removing the redundancy.

Comment (olegbespalov) [reply grpc/metrics.go:10]: It could contain the module metrics, not only the stream's

Comment (codebien) [reply grpc/metrics.go:10]: Oh, that's good info then I think I would like `instanceMetrics`. Anyway, the name of the struct is minor, I would prefer instead to drop the redundancy from the fields.

```suggestion
type instanceMetrics struct {
	Streams                 *metrics.Metric
	StreamsMessagesSent     *metrics.Metric
	StreamsMessagesReceived *metrics.Metric
}
```

**Comment by olegbespalov** [comment]
>I'm fine with it, but what does it include for a stream? Does it contain both or only the sent messages?

It's how it was implemented where the metrics emitted using the GRCP stat. Why is generally raising that after considering this, I concluded that streams are still requests, so maybe it's not a bug, but a feature. So that's why would like to discuss. Let's try to put it that way. Why do you think that streams should not be there? :sweat_smile: 

>I would prefer a minimal approach because metrics could be difficult to remove. So I would prefer if people asks to add them specifically. It would be difficult to get a negative feedback around it. No-one will tell us what they do not use, instead, it is easier to get a request to add a new one.

Not sure how properly interpret that. It could also sound like we don't need any metrics till the customers raise the issues, in other words, are you good with emitting at least these 3 new metrics? :thinking:

**Comment by codebien** [comment]
> Why do you think that streams should not be there?

My feeling is that it could be useful to distinguish between received and sent, so in case of troubleshooting of my service, I can understand if I have an issue with the load generator or on the SUT.

This is mostly guessing because I'm not very expert on this user journey.

>at least these 3 new metrics?

Sure, I'm for minimal not zero, and we need an opinionated proposal in some sense. Also, metrics are the entry point for checking that things are working correctly, so we need at least some essential set because running a test with zero metrics could be useless.

File: grpc/stream.go [hunk 2]
```diff
@@ -127,7 +140,19 @@ func (s *stream) loop() {
...
 	s.tq.Queue(func() error {
+		metrics.PushIfNotDone(s.vu.Context(), s.vu.State().Samples, metrics.Sample{
+			TimeSeries: metrics.TimeSeries{
+				Metric: s.instanceMetrics.GRPCStreamsMessagesReceived,
+				Tags:   s.tagsAndMeta.Tags,
+			},
+			Time:     time,
+			Metadata: s.tagsAndMeta.Metadata,
+			Value:    1,
+		})
```
Comment (mstoykov) [inline grpc/stream.go:154]: nit; I would argue this should be outside of the event loop.

And if it isn't, shouldn't `time.Now` call be inside teh event loop as well ?

**Review by mstoykov — APPROVED** [review]
LGTM! 

A single nit that isn't blockign

File: grpc/stream.go (hunk 2 above)
Comment (olegbespalov) [reply grpc/stream.go:154]: right, I'll adjust :+1:

File: grpc/metrics.go (hunk 1 above)
Comment (codebien) [reply grpc/metrics.go]: It sounds we haven't removed the GRPC redundancy from the fields
